Automattic/newspack-plugin
# feat(memberships): content gate page

### All Submissions:






### Changes proposed in this Pull Request:

Implements a custom gate solution on top of Woo Memberships integrated with Newspacks' reader activation system.

This PR covers the following:
 - Add a "Memberships Integration" setting in the RAS settings wizard
 - Implement an editor with custom settings for a Content Gate
 - Suppress Newspack Campaigns prompts on gated content (with a notification informing the publisher about it)
 - Reload the page once RAS detects a new reader so it unlocks the content

### RAS Settings

<img width="1114" alt="image" src="https://user-images.githubusercontent.com/820752/227539466-d30ee85b-114a-42c3-a431-811f5c70533a.png">

### Editor

<img width="1412" alt="image" src="https://user-images.githubusercontent.com/820752/227539602-feff38bc-5a80-4173-a133-19f69896c79c.png">

### Gate settings

<img width="280" alt="image" src="https://user-images.githubusercontent.com/820752/227543167-1e8a4ebe-f556-4f75-8ae9-c7147d07bf98.png">


### Gate flow with reader registration

https://user-images.githubusercontent.com/820752/227540775-b91c5744-9294-4e5a-994b-cb560102919e.mov

### Gate flow with donation subscription

https://user-images.githubusercontent.com/820752/227541516-9c884bb4-7389-4709-9af9-2ec31dff9a16.mov

The next steps for this feature will be:
 - Implement the "overlay" style with additional styling options
 - Implement the "Product Purchase Button Block" to support other subscription products as part of the gate
 - Support some type of "purchase" event in RAS, so it can reload the page on that occasion and not only the "reader" event
 - An automated initial draft for the gate, based on memberships plans configuration
 - Patterns for a landing page to offer the plans

### How to test the changes in this Pull Request:

1. Check out this branch, make sure you have:
    1. Woo Memberships active 
    2. A registration-based plan that locks posts
    3. RAS active
4. Navigate anonymously around posts and confirm the plugin's standard lock is in place
5. Go to "Newspack -> Engagement" and click to "Configure" the "Content Gate"
6. Confirm the gate starts as a draft with a single paragraph
7. Place a reader registration block and publish the gate
8. Visit a post anonymously and confirm the gate renders
9. Register the reader and confirm the page refreshes after 2 seconds and the content is unlocked
10. Remove the registration-based plan and create a plan based on subscription length tied to the Donate: Monthly and Donate: Yearly products
11. Edit the gate and change the reader registration block to a donate block
12. Visit a post anonymously, go through the donation flow, and confirm the page unlocks

#### Custom settings

In the content gate editor:

1. Toggle off the "Apply fade to last paragraph", visit a locked post and confirm the gradient is no longer applied
2. Change the paragraph count, visit a locked post and confirm it's applied
3. Edit a locked post, add a "More Block", visit and confirm the gate is now placed where the block was placed
4. Edit the gate, toggle off "Use “More” tag to manually place content gate", and confirm it's no longer using the tag for defining the gate placement

#### Campaigns integration

1. With the Newspack Campaigns plugin active, confirm you see the notice in the gate editor: "Newspack Campaigns: Prompts will not be displayed when rendering a gated content."
2. Add an overlay prompt that should be rendered inside a post
3. Visit the gated post and confirm the prompt does not render
4. Disable the membership plan, visit the post again (fresh session, if needed) and confirm the prompt renders
5. Disable the Newspack Campaigns plugin and confirm the notice does not render in the gate editor

### Other information:
